### PR TITLE
In-memory engine: fix data inconsistency due to unflushed data that exists when schedule load

### LINKED ISSUE #17645: In-memory Engine: data inconsistency in table: xx, index: yy, index-count:50 != record-count:49
## Bug Report

```log
[2024/10/12 17:16:22.112 +00:00] [ERROR] [reporter.go:231] ["indexLookup found data inconsistency"] [conn=2743074836] [session_alias=] [table_name=xx] [index_name=yy] [index_cnt=50] [table_cnt=49] ...
\"key\":\"74800000000000006C5F7203800BCB18172332B1038000002E7022329F\",\"mvcc\":{\"info\":{\"writes\":[{\"type\":1,\"start_ts\":453182325394243638,\"commit_ts\":453182325394243643},{\"start_ts\":453182325394243629,\"commit_ts\":453182325394243634},{\"start_ts\":453182325394243621,\"commit_ts\":453182325394243624},{\"start_ts\":453182325394243615,\"commit_ts\":453182325394243616},{\"start_ts\":453182325394243605,\"commit_ts\":453182325394243606},{\"start_ts\":453182325394243600,\"commit_ts\":453182325394243603},{\"start_ts\":453182325394243593,\"commit_ts\":453182325394243599},{\"start_ts\":453182325381136455,\"commit_ts\":453182325394243587},{\"start_ts\":453182325381136446,\"commit_ts\":453182325381136447},{\"start_ts\":453182325381136432,\"commit_ts\":453182325381136435},{\"start_ts\":453182325381136418,\"commit_ts\":453182325381136421},{\"type\":1,\"start_ts\":453181773711933489,\"commit_ts\":453181773711933493}],
```

The error message indicates that TiDB read a valid index value from the coprocessor, but the corresponding row value was not found in a subsequent request within the same transaction.

### What version of TiKV are you using?

v8.3.0

### Steps to reproduce

```bash
mysql --comments -h $HOST -P $PORT -u root -Ee 'DELETE FROM xx where id % 10000 > 10' \
  && sysbench --mysql-host=$HOST --mysql-user=root --mysql-port=$PORT --threads=10  --db-driver=mysql --report-interval=1 --mysql-db=zz --time=3600000  --number_workspaces=4000 --jobs_per_org=10000 --select_percentage=10 --cleanup_on_start=false many_delete_mvcc.lua run \
  || mysql --comments -h $HOST -P $PORT -u root -Ee 'update mysql.tidb set VARIABLE_VALUE="480h" where VARIABLE_NAME="tikv_gc_life_time";' ; date
```

### What did you expect?

No such error.

### What did happened?

sysbench fails.

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #17645 

What's Changed:

```commit-message
fix data inconsistency due to unflushed data that exists when schedule load
```
The root cause of the data inconsistency in the issue is that we schedule a load with unflushed data so that the unflushed data is neither planned to be written in IME nor loaded in IME from snapshot.
It is caused by the following sequence:
1. thread1: handle Msg::Apply
2. thread1: prepare_for_region, region 1 is not pending/loading/cached
3. thread1: write keys (saved unapplied cmds) without considering written to IME
4. thread1: handle anyother other Msg
5. thread2: `load_region` region 1 (now, region 1 becomes pending)
6. thread1: handle Msg::Apply
7. thread1: prepare_for_region, region 1 is pending and schedule real load

So, we have unflusehd data from line 5.
This PR fixes this by recording the region status of the first time calling prepare_for_region so that during the process of PollHandler::begin and PollHandler::end, the region status will not be changed.

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/in_memory_engine/src/write_batch.rs [hunk 1]
```diff
@@ -66,6 +67,14 @@ pub struct RegionCacheWriteBatch {
 
     // record the total durations of the prepare work for write in the write batch
     prepare_for_write_duration: Duration,
+
+    // Now, we have an assumption that in one round of batch system process, although the same
+    // region can call `prepare_for_region` multiple times, it can only call sequentially. This is
+    // say, we will not have this: prepare_for_region(region1), prepare_for_region(region2),
+    // prepare_for_region(region1).  In case to avoid this asssumption being broken, we record the
+    // regions that have called prepare_for_region and ensure that if the region is not the
+    // `currnet_region`, it is not recorded in this vec.
```
Comment (overvenus) [inline components/in_memory_engine/src/write_batch.rs:76]: Could you elaborate the definition of "one round of batch system process"?

---

I'm not sure if this is relevant, but an apply fsm may be put back to "batch" immediately and poll (`handle_normal`) again if new messages arrives after `handle_nromal` and before `schedule`.

https://github.com/tikv/tikv/blob/0b1124b92896b48fdd8290a20e55246cf1299fe3/components/batch-system/src/batch.rs#L408-L447

Comment (SpadeA-Tang) [reply components/in_memory_engine/src/write_batch.rs]: Im not sure what you mean of "poll again if new messages arrives".

File: components/in_memory_engine/src/write_batch.rs [hunk 2]
```diff
@@ -490,9 +491,7 @@ impl WriteBatch for RegionCacheWriteBatch {
         }
         let time = Instant::now();
         // verify that the region is not prepared before
-        self.prepared_regions
-            .iter()
-            .for_each(|id| assert_ne!(*id, region.id));
+        assert!(!self.prepared_regions.iter().any(|id| *id == region.id));
```
Comment (overvenus) [inline components/in_memory_engine/src/write_batch.rs:494]: Please print region ID if the assert fails.

File: components/in_memory_engine/src/write_batch.rs [hunk 3]
```diff
@@ -11,7 +11,11 @@ use engine_traits::{
     CacheRegion, EvictReason, MiscExt, Mutable, RegionCacheEngine, Result, WriteBatch,
     WriteBatchExt, WriteOptions, CF_DEFAULT,
 };
+<<<<<<< HEAD
+use smallvec::SmallVec;
+=======
 use kvproto::metapb;
+>>>>>>> master
```
Comment (overvenus) [inline components/in_memory_engine/src/write_batch.rs:18]: Merge conflict.

File: components/in_memory_engine/src/write_batch.rs [hunk 4]
```diff
@@ -472,11 +485,24 @@ impl WriteBatch for RegionCacheWriteBatch {
     }
 
     fn prepare_for_region(&mut self, region: &metapb::Region) {
+        let cached_region = CacheRegion::from_region(region);
+        if let Some(current_region) = &self.current_region
+            && current_region == &cached_region
```
Comment (glorv) [inline components/in_memory_engine/src/write_batch.rs:490]: Is it possible that epoch version changes between this two prepare? I think maybe check the region id is enough

File: components/in_memory_engine/src/write_batch.rs [hunk 5]
```diff
@@ -472,11 +485,24 @@ impl WriteBatch for RegionCacheWriteBatch {
...
+        {
+            return;
+        }
         let time = Instant::now();
+        // verify that the region is not prepared before
+        if let Some(region_id) = self.prepared_regions.iter().find(|id| **id == region.id) {
```
Comment (glorv) [inline components/in_memory_engine/src/write_batch.rs:496]:
```suggestion
        if self.prepared_regions.contains(&region.id) {
```

File: components/in_memory_engine/src/write_batch.rs (hunk 4 above)
Comment (SpadeA-Tang) [reply components/in_memory_engine/src/write_batch.rs]: Good catch

**Review by glorv — APPROVED** [review]

File: components/collections/src/lib.rs [hunk 6]
```diff
@@ -10,6 +10,10 @@ pub type HashMap<K, V> =
 pub type HashSet<T> = std::collections::HashSet<T, std::hash::BuildHasherDefault<fxhash::FxHasher>>;
 pub use std::collections::hash_map::Entry as HashMapEntry;
 
+pub fn hash_map_with_capacity<K: Hash + Eq, V>(capacity: usize) -> HashMap<K, V> {
```
Comment (overvenus) [inline components/collections/src/lib.rs:13]: Dead code?

File: components/in_memory_engine/src/write_batch.rs [hunk 7]
```diff
@@ -472,7 +485,20 @@ impl WriteBatch for RegionCacheWriteBatch {
     }
 
     fn prepare_for_region(&mut self, region: &metapb::Region) {
+        if let Some(current_region) = &self.current_region
+            && current_region.id == region.id
+        {
+            return;
+        }
```
Comment (overvenus) [inline components/in_memory_engine/src/write_batch.rs:492]: Please add some comments about why the check is needed.

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: components/collections/src/lib.rs (hunk 6 above)
Comment (SpadeA-Tang) [reply components/collections/src/lib.rs:13]: Yes, I forget to remove it.

File: components/in_memory_engine/src/write_batch.rs (hunk 7 above)
Comment (SpadeA-Tang) [reply components/in_memory_engine/src/write_batch.rs:494]: Done

**Review by overvenus — APPROVED** [review]

**Pull request merged**